src/cio_websocket_location_handler.c

```diff
@@ -0,0 +1,297 @@
+	} else {
+		return CIO_HTTP_CB_ERROR;
+	}
+}
+
+static bool find_requested_sub_protocol(struct cio_websocket_location_handler *handler, const char *name, size_t length)
+{
+	for (unsigned int i = 0; i < handler->number_subprotocols; i++) {
+		const char *sub_protocol = handler->subprotocols[i];
+		size_t name_length = strlen(sub_protocol);
+		if (name_length == length) {
+			if (memcmp(sub_protocol, name, length) == 0) {
+				handler->chosen_subprotocol = i;
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+static void check_websocket_protocol(struct cio_websocket_location_handler *handler, const char *at, size_t length)
+{
+	if (handler->subprotocols != NULL) {
+		const char *start = at;
+		while (length > 0) {
+			if (!isspace(*start) && (*start != ',')) {
+				const char *end = start;
+				while (length > 0) {
+					if (*end == ',') {
+						ptrdiff_t len = end - start;
+						if (find_requested_sub_protocol(handler, start, len)) {
+							return;
+						}
+
+						start = end;
+						break;
+					}
+					end++;
+					length--;
+				}
+				if (length == 0) {
+					ptrdiff_t len = end - start;
+					if (find_requested_sub_protocol(handler, start, len)) {
+						return;
+					}
+				}
+			} else {
+				start++;
+				length--;
+			}
+		}
+	}
+}
+
+static enum cio_http_cb_return handle_field(struct cio_http_client *client, const char *at, size_t length)
+{
+	static const char sec_key[] = "Sec-WebSocket-Key";
+	static const char ws_version[] = "Sec-WebSocket-Version";
+	static const char ws_protocol[] = "Sec-WebSocket-Protocol";
+
+	struct cio_websocket_location_handler *ws = container_of(client->handler, struct cio_websocket_location_handler, http_location);
+
+	if ((sizeof(sec_key) - 1 == length) && (cio_strncasecmp(at, sec_key, length) == 0)) {
+		ws->flags.current_header_field = CIO_WS_HEADER_SEC_WEBSOCKET_KEY;
+	} else if ((sizeof(ws_version) - 1 == length) && (cio_strncasecmp(at, ws_version, length) == 0)) {
+		ws->flags.current_header_field = CIO_WS_HEADER_SEC_WEBSOCKET_VERSION;
+	} else if ((sizeof(ws_protocol) - 1 == length) && (cio_strncasecmp(at, ws_protocol, length) == 0)) {
+		ws->flags.current_header_field = CIO_WS_HEADER_SEC_WEBSOCKET_PROTOCOL;
+	}
+
+	return CIO_HTTP_CB_SUCCESS;
+}
+
+static enum cio_http_cb_return handle_value(struct cio_http_client *client, const char *at, size_t length)
+{
+	enum cio_http_cb_return ret;
+
+	struct cio_websocket_location_handler *ws = container_of(client->handler, struct cio_websocket_location_handler, http_location);
+
+	switch (ws->flags.current_header_field) {
+	case CIO_WS_HEADER_SEC_WEBSOCKET_KEY:
+		ret = save_websocket_key(ws->sec_web_socket_key, at, length);
+		break;
+
+	case CIO_WS_HEADER_SEC_WEBSOCKET_VERSION:
+		ret = check_websocket_version(at, length);
+		break;
+
+	case CIO_WS_HEADER_SEC_WEBSOCKET_PROTOCOL:
+		ws->flags.subprotocol_requested = 1;
+		ret = CIO_HTTP_CB_SUCCESS;
+		check_websocket_protocol(ws, at, length);
+		break;
+
+	case CIO_WS_HEADER_UNKNOWN:
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this redundant "case" clause. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3AS3458)